Feature 1

simt11 opened the pull request.

simt11 committed:
Feature_1

1) Инициализирую retrofit
2) Добавляю интерфейс, который определяет как retrofit общается с веб-сервером с помощью HTTP-запросов.(Часть url динамическая)
3) Ленивая инициализация retrofitService, чтобы убедиться, что она инициализируется при первом использовании.

4) Создаю BinViewModel

simt11 committed:
5) Добавил в ActivityMain контейнер для фрагментов.
6) Создал стартовый фрагмент
7) Установил автопереход на него.
8) Создал Fun.kt для новых функций.
9) Добавил функцию переход во фрагмент для MainActivity

simt11 committed:
10)Добавил компактный AppTextWatcher
11)

simt11 committed:
10)Добавил компактный AppTextWatcher
11)

simt11 committed:
11)

simt11 committed:
11) Изменен тип конвертера на JSON
12) Создан класс BinCard со всеми параметрами карты
13) String by mutableStateOf("") - который работал бы с Compose. Заменено на LiveData, для работы с binding
14) Создана разметка для отображения данных (не доделана)
15) Инициализация viewModel теперь идет через ViewModelProvider.
16) А вывод данных из корутин через observe для LiveData

simt11 committed:
17) Добавлены переходы по ссылке, телефону и координаты в смежные приложения

simt11 committed:
17) Добавлены переходы по ссылке, телефону и координаты в смежные приложения
18) Интенты вынесены в отдельные функции в Fun
19) Добавлены функции показа и скрытия клавиатуры.
20) Реализовано сохранение номера карты через SharedPreferences.
21) Слушатель набора номера заменен на кнопку. Чтобы не было нагрузки, из-за частого обновления.
22) Добавлено скрытие клавиатуры при нажатие на экран помимо EditText

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+package com.example.bininfo
```
есть два способа структурирования исходного кода:
- группировать по фичам (то есть по мейн скрину все в одну папку)
- или группировать по типам
Я вижу что у тебя сделан второй вариант, но не полностью. Вот эта модель, это ответ от сетевого запроса, а значит должна лежать в пакете network, а лежит просо в корне проекта. Так же в корне лежит MainActivity, а она должна лежать скорее всего в ui/screens. Пробегись по всем остальным файлам и проведи ревизию структуры

simt11 replied:
Исправил

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+
+@Serializable
+data class BinCard(
+    val number: Number? = null,
```
здесь не нужно у каждого параметра писать  = null, сериалайзер сам подставит нули, если данные не придут. ППредлагаю убрать  = null, тк это засоряет код и ухудшает читаемость

simt11 replied:
Так все равно приложение крашится

fantasianna commented on app/src/main/java/com/example/bininfo/ui/screens/HomeFragment.kt:
```diff
@@ -0,0 +1,111 @@
+val BIN_NUMBER: String = ""
+private const val MIN_LENGTH_CARD = 4
+
+class HomeFragment(context: Context) : Fragment(R.layout.fragment_home) {
```
если этот фрагмент использует BinViewModel и наверное используется именно для отрисовки БИНов, то имеет смыл его назвать так же как и вью модель, то есть BinFragment, чтобы глядя на на файлы фрагмент и вью модели была сразу очевидна их кармическая связь

simt11 replied:
Исправил

fantasianna commented on app/src/main/java/com/example/bininfo/MainActivity.kt:
```diff
@@ -2,10 +2,21 @@ package com.example.bininfo
+
+    override fun onStart() {
+        super.onStart()
+        replaceFragment(HomeFragment(this), false)
```
зачем тут фрагмент внутри активити? это имело бы смысл, если бы у тебя была смена фрагментов, как в приложении телеграмма. А если у тебя один экран, то это скользкий момент, получается оверинжиниринг в моменте, где по факту это не нужно. Предлагаю избавиться от вложенности фрагмента в активити, и сделать все внутри активити

simt11 replied:
Здесь подразумевается второй фрагмент с историей запросов. Для этого вынес из активити. (Плюс так по привычке сделал :)

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+    class Number(
+        val length: Int? = null,
+        val luhn: Boolean? = null,
+    ) {}
```
{} - не нужно, удали везде где нет тела

simt11 replied:
Сделано

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+        val name: String? = null,
+        val emoji: String? = null,
+        val currency: String? = null,
+        val latitude: Int? = null,
```
а широта и долгота правда приходят интами? 
я думала у них такой формат: 41.40338, 2.17403

simt11 replied:
Да, ты права.

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+    val brand: String? = null,
+    val prepaid: Boolean? = null,
+    val country: Country? = null,
+    val bank: Bank? = null,
```
очень подозрительно выглядят контракты, в которых все параметры опциональные? получается вообще никакого контракта взаимодействия с сервером нет и мы не можем быть уверены в приходе вообще никаких парамметров. Обычно контракты составляются таким образом, что у каждого параметра написано, опциональный он или нет, то есть предусматривает ли контракт, что в некоторых случаях параметр не придет или придет со значением null. Я так понимаю что тут нет описания контракта, но все равно делать все параметры опциональными не стоит. Посмотри для разных входных данных, какие параметры могут не прийти, и только их сделай опциональными, а остальные - обязательными

fantasianna replied:
то же актуально для всех моделей ниже

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+    val scheme: String? = null,
+    val type: String? = null,
+    val brand: String? = null,
+    val prepaid: Boolean? = null,
```
булевы параметры надо начинать с is или has, то есть тут будет isPrepaid, проверь для всех отсальных

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+data class BinCard(
+    val number: Number? = null,
+    val scheme: String? = null,
+    val type: String? = null,
```
тут набор типов не ограничен какими-то значениями? можно сделать енам?

fantasianna replied:
не крит, просто хорошо бы понимать

fantasianna commented on app/src/main/java/com/example/bininfo/BinCard.kt:
```diff
@@ -0,0 +1,39 @@
+    @Serializable
+    class Number(
+        val length: Int? = null,
+        val luhn: Boolean? = null,
```
это алгоритм Луны?))
кажется читаемее было бы значение isLuhnAlgorithm

simt11 replied:
Я правильно понял, что здесь названия должны полностью совпадать с названиями JSON ответа? А так согласен. Твой вариант более читаемый

fantasianna replied:
да должен совпадать с джейсоном, так что пусть останется как есть

fantasianna commented on app/src/main/java/com/example/bininfo/network/BinApiService.kt:
```diff
@@ -0,0 +1,27 @@
+
+interface BinApiService {
+    @GET("{BIN}")
+    suspend fun getBinInfo(@Path("BIN", encoded = true) BIN: String): BinCard
```
параметры методов называем маленькими буквами, тут должно быть bin: String

simt11 replied:
Исправил

fantasianna commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,19 +1,16 @@
-        app:layout_constraintEnd_toEndOf="parent"
-        app:layout_constraintStart_toStartOf="parent"
-        app:layout_constraintTop_toTopOf="parent" />
+<androidx.constraintlayout.widget.ConstraintLayout
```
не нужно делать вложенные друг в друга ConstraintLayout, они созданы как раз для того, чтобы избежать вложенности внутри себя.
Ты можешь добавить android:id="@+id/dataContainer" к констрейнту верхнего уровня

simt11 replied:
Исправил

fantasianna commented on app/src/main/res/layout/fragment_home.xml:
```diff
@@ -0,0 +1,215 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout
```
попробуй запустить свой проект на эмуляторе с маленьким экраном. Насколько я знаю ConstraintLayout не умеет скролится, и если у тебя будет маленький экран, то весь контент на него не влезет и пользователь не увидит низ списка и не сможет просколить вниз. Быстрое решение: поместить ConstraintLayout внутрь ScrollView. Но идеальное решение, это использовать RecyclerView. Этот контейнер используется для 90% экранов на андроиде и с ним тебе все равно придется познакомиться. 
https://habr.com/ru/post/705064/

simt11 replied:
Эмулятор с маленьким экраном, не запускаетя(

fantasianna replied:
это критическое замечание, надо разобраться, попробуй другое устройство в эмуляторе

fantasianna replied:
подумала тут RecyclerView не подойдет, потому что экран формируется статически, нет профита от RecyclerView. но в ScrollView надо завернуть твой ConstraintLayout и провериь на маленьком экране. Так же это можно проверить, если открыть клавиатуру и посмотреть, можешь ли ты проскролить экран вниз

simt11 replied:
Разобрался с проблемами эмулятора. Оказалось, что закончилось место на диске. Как увеличил, так и заработало)))))

simt11 replied:
По поводу скролинга, буду еще разбираться

simt11 replied:
Добавил скролинг во фрагмент.

fantasianna commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,17 @@
     <string name="app_name">BinInfo</string>
+    <string name="enter_card_number">Введите первые 8 цифр номера карты</string>
+    <!-- TODO: Remove or change this placeholder text -->
+    <string name="hello_blank_fragment">Hello blank fragment</string>
```
это текст где-то используется?

simt11 replied:
Убрал этот текст

fantasianna commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,35 @@
+        <item name="android:textSize">18sp</item>
+    </style>
+
+    <style name="text_param">
```
ты выше использвал камел кейс для именования стилей, тут туже надо продолжить, будет textParam

simt11 replied:
Исправил на камел кейс

fantasianna commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,35 @@
+        <item name="android:layout_width">match_parent</item>
+    </style>
+
+    <style name="answer_param">
```
и тут

fantasianna commented on app/src/main/java/com/example/bininfo/utilits/appTextWatcher.kt:
```diff
@@ -0,0 +1,14 @@
+import android.text.Editable
+import android.text.TextWatcher
+
+class appTextWatcher(val onSuccess:(Editable?) -> Unit):TextWatcher {
```
я чаше встречала названия SimpleTextWatcher для таких реализаций, когда реализуется какая то часть интерфейса. Имена классов всегда начинаются с большой буквы, мы относимся к ним уважительно

fantasianna replied:
:TextWatcher нужен пробел между

simt11 replied:
Отметил себе по названиям. Класс удален, за ненадобностью. Теперь проверка происходит по нажатию

fantasianna commented on app/src/main/java/com/example/bininfo/utilits/Fun.kt:
```diff
@@ -0,0 +1,67 @@
+package com.example.bininfo.utilits
```
этот класс содержит экстеншены, я часто встречала для такого класса название Extensions.kt

simt11 replied:
я не понял, для какого класса(((( Или ты имела ввиду название папки?

simt11 replied:
Изменил Fun на Extensions

fantasianna replied:
я имела ввиду название файла

fantasianna commented on app/src/main/java/com/example/bininfo/utilits/Fun.kt:
```diff
@@ -0,0 +1,67 @@
+
+fun Fragment.intentPhone(Phone: String) {
+    val query: Uri = Uri.parse("tel:${Phone}")
+    val intent = Intent(Intent.ACTION_DIAL, query)
```
а пермишены в манифесте объявлены для этого акшена?

simt11 replied:
не объявлены... 
Как я понимаю, здесь не нужны разрешения дополнительные. Потому, что это не звонок по телефону, а только передача номера в звонилку. Фактически приложение не звонит и разрешение поэтому ему не нужно.

fantasianna replied:
ага поняла хорошо

fantasianna commented on app/src/main/java/com/example/bininfo/ui/screens/HomeFragment.kt:
```diff
@@ -0,0 +1,111 @@
+import com.example.bininfo.BinCard
+import com.example.bininfo.R
+import com.example.bininfo.databinding.FragmentHomeBinding
+import com.example.bininfo.utilits.hideKeyboard
```
а где showKeyboard? он не используется в итоге? может удалить его тогда из екстеншенов?

simt11 replied:
да, ок)

simt11 committed:
Внесены изменения по замечаниям в пулреквесте

simt11 committed:
Внесены изменения по замечаниям в пулреквесте.

1) Фрагмент начального экрана обернут в Скролинг.
2) Дабавлен переход на фрагмент истории запросов.

fantasianna commented on app/src/main/java/com/example/bininfo/ui/screens/BinFragment.kt:
```diff
@@ -0,0 +1,115 @@
+            val string = binding.enterCardNumber.text.toString()
+            if (string.length > MIN_LENGTH_CARD) {
+                hideKeyboard()
+                viewModel.getBinInfo(string)
```
вот это место надо будет переделать, чтобы логику убрать во вью модель, а фрагмент оставить максимально глупым и чтобы он только умел отрисовывать. Сядем вместе потом, когда будет готов поиск

fantasianna commented on app/src/main/java/com/example/bininfo/ui/screens/BinFragment.kt:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        binding.answerCoordinates.setOnClickListener {
+            val Coordinates = binding.answerCoordinates.text.toString()
```
названия переменных всегда с маленькой буквы

simt11 replied:
Точно. Копипастил

fantasianna commented on app/src/main/java/com/example/bininfo/ui/screens/BinFragment.kt:
```diff
@@ -0,0 +1,115 @@
+        binding.answerBankPhone.setOnClickListener {
+            val bankPhone = binding.answerBankPhone.text.toString()
+            if (bankPhone.isNotEmpty()) {
+                intentPhone(bankPhone)
```
логику роутинга тоже надо будет убрать во вью модель, фргамент не должен принимать таких важных решений, в нем вообще не должно быть какой либо логики, кроме отрисовки

fantasianna commented on app/src/main/java/com/example/bininfo/ui/screens/BinFragment.kt:
```diff
@@ -0,0 +1,115 @@
+
+    override fun onResume() {
+        super.onResume()
+        if (preferences.contains(BIN_NUMBER)) {
```
логику работы с преференсами надо тоже убрать во вью модель
